tensorflow/core/kernels/ragged_fill_empty_rows_op_gpu.cu.cc

```diff
@@ -0,0 +1,598 @@
+    GpuLaunchConfig cfg, Tindex nrows, const Tindex* indices,
+    Tindex* elements_per_row, int* rows_are_not_ordered,
+    int* first_invalid_index) {
+  GPU_1D_KERNEL_LOOP(i, cfg.virtual_thread_count) {
+    Tindex row = indices[i];
+    if (row < 0 || row >= nrows) {
+      GpuAtomicMin(first_invalid_index, i);
+      continue;
+    }
+    GpuAtomicAdd(&elements_per_row[row], 1);
+    // Note that this only needs to compare rows, not columns, to satisfy the
+    // row-major order invariant.
+    if (i > 0 && row < indices[(i - 1)]) {
+      GpuAtomicMax(rows_are_not_ordered, 1);
+    }
+  }
+}
+
+template <typename Tindex>
+__global__ __launch_bounds__(1024) void CopyRowIndicesKernel(
+    GpuLaunchConfig cfg, const Tindex* indices, Tindex* row_indices) {
+  GPU_1D_KERNEL_LOOP(i, cfg.virtual_thread_count) {
+    row_indices[i] = indices[i];
+  }
+}
+
+// Sets empty_row_indicator[row] to whether the row is empty.
+template <typename Tindex>
+__global__ __launch_bounds__(1024) void ComputeEmptyRowIndicatorKernel(
+    GpuLaunchConfig cfg, const Tindex* elements_per_row,
+    bool* empty_row_indicator) {
+  GPU_1D_KERNEL_LOOP(row, cfg.virtual_thread_count) {
+    empty_row_indicator[row] = elements_per_row[row] == 0;
+  }
+}
+
+// Copies indices and values to output_indices and output_values, leaving space
+// in the output for the new elements that will be inserted wherever there is an
+// empty row.
+template <typename T, typename Tindex>
+__global__ __launch_bounds__(1024) void ScatterInputElementsKernel(
+    GpuLaunchConfig cfg, Tindex nrows,
+    const Tindex* input_index_map, const Tindex* indices, const T* values,
+    const Tindex* num_new_rows_before, Tindex* output_indices, T* output_values,
+    Tindex* reverse_index_map) {
+  GPU_1D_KERNEL_LOOP(i, cfg.virtual_thread_count) {
+    Tindex input_i = input_index_map ? input_index_map[i] : i;
+    Tindex row = indices[input_i];
+    Tindex output_i = i + num_new_rows_before[row];
+    output_indices[output_i] = indices[input_i];
+    output_values[output_i] = values[input_i];
+    if (reverse_index_map) {
+      reverse_index_map[input_i] = output_i;
+    }
+  }
+}
+
+// Sets the new elements (which correspond to the empty rows in the
+// input) in output_indices and output_values.
+template <typename T, typename Tindex>
+__global__ __launch_bounds__(1024) void ScatterNewElementsKernel(
+    GpuLaunchConfig cfg, const T* default_value,
+    const Tindex* num_new_rows_through, const Tindex* input_row_ends,
+    const bool* empty_row_indicator, Tindex* output_indices, T* output_values) {
+  GPU_1D_KERNEL_LOOP(row, cfg.virtual_thread_count) {
+    if (!empty_row_indicator[row]) continue;  // Only process empty rows
+    Tindex input_i = (row == 0 ? 0 : input_row_ends[row - 1]);
+    Tindex output_i = input_i + (row == 0 ? 0 : num_new_rows_through[row - 1]);
+    output_indices[output_i] = row;
+    output_values[output_i] = *default_value;
+  }
+}
+
+}  // namespace
```

Comment:
Is it possible to share these lines with `sparse_fill_empty_rows_op_gpu.cu.cc` by passing rank = 1? We could move the common functions into an util file to reduce code duplication.